build-support/bin/terraform_tool_versions.py

```diff
@@ -0,0 +1,266 @@
+        for li in items
+        if not li.a.get("href").startswith("..")
+    ]
+    return links
+
+
+@dataclass(frozen=True)
+class TFVersionLinks:
+    platform_links: Links
+    sha256sums_link: Link
+    signature_link: Link
+
+
+def get_info_for_version(url) -> TFVersionLinks:
+    """Get list of binaries and signatures for a version of Terraform."""
+    logging.info(f"getting info for version at {url}")
+    all_links = get_tf_links(get_tf_page(url))
+    signature_links, platform_links = partition(lambda i: "SHA" in i.link, all_links)
+
+    def link_ends_with(what: str) -> Link:
+        return next(filter(lambda s: s.link.endswith(what), signature_links))
+
+    return TFVersionLinks(
+        platform_links,
+        sha256sums_link=link_ends_with("SHA256SUMS"),
+        signature_link=link_ends_with("SHA256SUMS.sig"),
+    )
+
+
+@dataclass(frozen=True)
+class VersionHash:
+    filename: str
+    sha256sum: str
+
+
+@dataclass(frozen=True)
+class VersionHashes:
+    sha256sums: List[VersionHash]
+    signature: bytes
+
+    def by_file(self) -> Dict[str, str]:
+        """Get sha256sum by filename."""
+        return {x.filename: x.sha256sum for x in self.sha256sums}
+
+
+def parse_sha256sums_file(file_text: str) -> List[VersionHash]:
+    """Parse Terraform's sha256sums file."""
+    return [
+        VersionHash(**x)
+        for x in csv.DictReader(
+            StringIO(file_text),
+            delimiter=" ",
+            skipinitialspace=True,
+            fieldnames=["sha256sum", "filename"],
+        )
+    ]
+
+
+def parse_signatures(links: TFVersionLinks, verifier: GPGVerifier) -> VersionHashes:
+    """Parse and verify GPG signatures of SHA256SUMs."""
+
+    sha256sums_raw = requests.get(links.sha256sums_link.link)
+    sha256sums = parse_sha256sums_file(sha256sums_raw.text)
+
+    signature = requests.get(links.signature_link.link).content
+
+    vr = verifier.validate_signature(signature, sha256sums_raw.content)
+    if not vr.valid:
+        logging.error(f"signature is not valid for {links.signature_link.text}")
+        raise RuntimeError("signature is not valid")
+    else:
+        logging.info(f"signature is valid for {links.signature_link.text}")
+
+    return VersionHashes(sha256sums, signature)
+
+
+def get_file_size(url) -> int:
+    """Get content-length of a binary."""
+    logging.info(f"fetching content-length for {url}")
+    r = requests.head(url)
+    return int(r.headers["content-length"])
+
+
+def parse_download_url(url: str) -> Tuple[str, str]:
+    """Get the version and platform from the url."""
+    filename = Path(urlparse(url).path).stem
+    _, version, platform_name, platform_arch = filename.split("_")
+    return version, platform_name + "_" + platform_arch
+
+
+def is_prerelease(version_slug: str) -> bool:
+    """Determine if a Terraform version is a prerelease version (alpha, beta, or rc)"""
+    return any((x in version_slug for x in {"alpha", "beta", "rc"}))
```

Comment:
Maybe it would be more robust to check for a regex match for 

`^terraform_\d+\.\d+.\d+$` rather than assume that these three strings are the only ones that may appear to denote a prerelease? If they suddenly introduce some other convention, it will defeat this logic, whereas they are very unlikely to introduce a release convention that isn't that regex.

Reply:
good point, and simple enough to implement